### PR TITLE
Added Pin Support, lock unlock, odometer and location for CAD

### PR DESCRIPTION
**Pull request by cdnninja**
Added pin support for lock unlock.  Testing still required.

### CODE REVIEW

**Comment by cdnninja** [comment]
@fuatakgun.  That is enough for one night.  So here is where this is at.  

- verify_pin method fails with a status code 200 back from the server.  I actually think I am messing up the token by passing the token within the method.   The json fails to parse the response. Could you take a look? 
- if I comment out the need for the verify_pin I the following.  Even with hard coded pin number. I suspect because I didn't validate first and pass that with the header to lock. 
- `2021-08-14 20:21:04 DEBUG (SyncWorker_2) [custom_components.kia_uvo.KiaUvoApiCA] kia_uvo - Received lock_action response {'error': {'errorCode': '7310', 'errorDesc': 'PIN incorrect.  Multiple incorrect attempts will require PIN reset'}, 'responseHeader': {'responseCode': 1, 'responseDesc': 'Failure'}}`  
- I added pin support to EU as well.  I wouldn't mind testing to make sure it doesn't break anything.  Not sure EU has value in PIN though. I didn't implement the usage just pass it to the class.

**Comment by cdnninja** [comment]
Well just kidding.  Removed the integration and set it up again and this all works great.   Not sure what changed.  Only thing that needs work is it doesn't refresh and validate it worked until next polling cycle. 

Could you test on the EU side I didn't break anything?  Then the question is how to make it so people can upgrade without uninstall re-install.

**Comment by cdnninja** [comment]
I attempted to get odometer working in CAD.   No errors occur but it doesn't display so I suspect I am not placing the value in the right spot.  Are you able to take a look?

**Comment by cdnninja** [comment]
@fuatakgun wanted to check in to see when you may be able to review?  Hoping to do more commits but wanted to get this big one in first.

**Comment by fuatakgun** [comment]
Thanks for great work, I am on vacation for next 2 weeks and I want to be sure that this is not breaking at least EU/Kia (my car) use cases before merging.

I have one general advice, can you squash all these changes into one commit? so in case anything goes wrong, rolling back can be easier and faster rather than finding correct spot.

**Comment by cdnninja** [comment]
Updated to a single check in.

File: custom_components/kia_uvo/KiaUvoApiCA.py [hunk 1]
```diff
@@ -21,9 +22,10 @@ def __init__(
         password: str,
         region: int,
         brand: int,
+        pin: int,
```
Comment (fuatakgun) [inline custom_components/kia_uvo/KiaUvoApiCA.py:25]: is this really integer? can it start with `zero`?

File: custom_components/kia_uvo/KiaUvoApiCA.py [hunk 2]
```diff
@@ -109,6 +112,19 @@ def get_cached_vehicle_status(self, token: Token):
         vehicle_status = {}
         vehicle_status["vehicleStatus"] = response
         vehicle_status["vehicleStatus"]["time"] = response["lastStatusDate"]
+        service_status = {}
+        service_status = self._get_service_status(token)
```
Comment (fuatakgun) [inline custom_components/kia_uvo/KiaUvoApiCA.py:116]: as of now, I am not following `_` (underscore naming pattern) ,you can use it without it.

File: custom_components/kia_uvo/KiaUvoApiCA.py [hunk 3]
```diff
@@ -51,6 +53,7 @@ def __init__(
     def login(self) -> Token:
         username = self.username
         password = self.password
+        pin = self.pin
```
Comment (fuatakgun) [inline custom_components/kia_uvo/KiaUvoApiCA.py:56]: you need to store this inside `token.set` as a last variable, which can be used for creating Token instance from json

File: custom_components/kia_uvo/KiaUvoApiEU.py [hunk 4]
```diff
@@ -28,8 +28,9 @@ def __init__(
         region: int,
         brand: int,
         use_email_with_geocode_api: bool = False,
+        pin: int,
```
Comment (fuatakgun) [inline custom_components/kia_uvo/KiaUvoApiEU.py:31]: no need to introduce this as of now, just set a default value

File: custom_components/kia_uvo/KiaUvoApiEU.py [hunk 5]
```diff
@@ -28,8 +28,9 @@ def __init__(
...
     ):
-        super().__init__(username, password, region, brand, use_email_with_geocode_api)
+        super().__init__(username, password, region, brand, use_email_with_geocode_api, pin)
```
Comment (fuatakgun) [inline custom_components/kia_uvo/KiaUvoApiEU.py:33]: keep the correct constructor signature, pin should come after `password` as it is in CA implementation

File: custom_components/kia_uvo/KiaUvoApiCA.py [hunk 6]
```diff
@@ -21,9 +22,10 @@ def __init__(
         password: str,
         region: int,
         brand: int,
+        pin: int,
         use_email_with_geocode_api: bool = False,
     ):
-        super().__init__(username, password, region, brand, use_email_with_geocode_api)
+        super().__init__(username, password, region, brand, pin, use_email_with_geocode_api)
```
Comment (fuatakgun) [inline custom_components/kia_uvo/KiaUvoApiCA.py:28]: can we move `pin` after `password`

File: custom_components/kia_uvo/Vehicle.py [hunk 7]
```diff
@@ -1,203 +1,195 @@
...
-from .const import *
+from .const import DOMAIN, BRANDS, BRAND_HYUNDAI, BRAND_KIA, DATE_FORMAT, VEHICLE_LOCK_ACTION
 from .KiaUvoApiImpl import KiaUvoApiImpl
 from .Token import Token
 
 _LOGGER = logging.getLogger(__name__)
 
-class Vehicle(object):
+
+class KiaUvoApiCA(KiaUvoApiImpl):
```
Comment (fuatakgun) [inline custom_components/kia_uvo/Vehicle.py:18]: I am lost here, why did you put this class into Vehicle.py ?

**Comment by fuatakgun** [comment]
hey, one question, if you want, I can create a new revision on this and we can review together. Is it fine for you?

**Comment by cdnninja** [comment]
Yes for sure.  I’m no developer so just hacking my way through it! 

One note I am hitting some bugs on refresh with my current code.  Looks like I’m maxing the API because of the location and pin check.

File: custom_components/kia_uvo/Vehicle.py (hunk 7 above)
Comment (cdnninja) [reply custom_components/kia_uvo/Vehicle.py:18]: Looks like a copy and paste mistake.  This was not what was running locally on my side.  Corrected now.

File: custom_components/kia_uvo/KiaUvoApiEU.py (hunk 5 above)
Comment (cdnninja) [reply custom_components/kia_uvo/KiaUvoApiEU.py]: Done, I know you are making changes anyway but good practice for me.

**Comment by cdnninja** [comment]
I fixed a few items and also gave you edit access.  So feel free to make commits against this pull request.  Or start a new one either way!

**Pull request closed without merging**